### PR TITLE
V2 - No more magic

### PR DESCRIPTION
**Pull request by ouchadam**
After using v1~ for over 2 years in multiple projects and the android landscape changing, the library needed some attention... V1 -> V2 is a massively breaking design change further reducing the scope of the library pushing more logic onto clients.

V2 aims to:

- Use a recognised update pattern.
- Reduce _Libraryside magic._
- Simplify the overall library. 

Removed :

```
LandingStrip.attach(viewPager)
```
```
app:tabLayoutId
```

`LandingStrip.setAdapter(landingStripAdapter)` replaces `attach` and the custom `tabLayoutId` by providing a more recognisable adapter pattern `createView/bindView` instead.

`LandingStripAdapter.notifyDatasetChanged` will drop, create and bind all tab views into the tab container.

`LandingStripAdapter.notifyItemChanged` will rebind a single tab without dropping the entire tab container from the view hierarchy.

```
LandingStrip.addPageOnChangeListener
```

`LandingStrip.addPageOnChangeListener` is no longer required since the support-v4 introduced `ViewPager.addPageOnChangeListener` instead of `ViewPager.setPageChangeListener`.   As a client of the library you're now required to pass the `LandingStrip` as a `PageChangeListener` to the view pager `viewPager.addOnPageChangeListener(landingStrip);`

```
DataSetObserver
```

`LandingStrip` no longer registers itself to a `PagerAdapter` and must instead have the set `LandingStripAdapter` notified of dataset changes.

```
TabOnClickListener
```

`LandingStrip` no longer automagically sets a click listener on the tab views, the client must do this manually **AND** call `LandingStrip.setCurrentItem(position)`, this is also required if and when calling `ViewPager.setCurrentItem(position)` manually.

```
Demos
```

All but the basic example demos have been removed, they'll be replaced with wiki pages if required.

Full example:

```java
static class TitledAdapter extends TabAdapter<TextView> {

    private final ViewPager viewPager;

    public TitledAdapter(LayoutInflater layoutInflater, int itemLayoutId, LandingStrip landingStrip, ViewPager viewPager) {
        super(layoutInflater, itemLayoutId, landingStrip, viewPager);
        this.viewPager = viewPager;
    }

    @Override
    protected void bindView(TextView view, int position) {
        super.bindView(view, position);
        bindPageTitleToView(view, position);
    }

    private void bindPageTitleToView(TextView itemView, int position) {
        PagerAdapter adapter = viewPager.getAdapter();
        if (adapter == null) {
            return;
        }
        itemView.setText(adapter.getPageTitle(position));
    }

}
```

Attach the `TabAdapter` to the `LandingStrip` and the `LandingStrip` to the `ViewPager`.

```java
LandingStrip landingStrip = (LandingStrip) findViewById(R.id.landing_strip);
TabAdapter tabAdapter = new TitledAdapter(
        getLayoutInflater(),
        R.layout.tab_simple_text,
        landingStrip,
        viewPager
);
landingStrip.setAdapter(tabAdapter);
viewPager.addOnPageChangeListener(landingStrip);
```

~Lots of debating~ Paired with @ataulm

### CODE REVIEW

File: core/src/main/java/com/novoda/landingstrip/Attributes.java [hunk 1]
```diff
@@ -35,28 +29,17 @@ static Attributes readAttributes(Context context, AttributeSet attrs) {
...
-
-            throwIfTabLayoutIsMissing(tabLayoutId);
-
             int indicatorColour = xml.getResourceId(R.styleable.LandingStrip_indicatorColor, DEFAULT_INDICATOR_COLOUR);
             int indicatorHeight = getDimensPixelSize(R.styleable.LandingStrip_indicatorHeight, DEFAULT_INDICATOR_HEIGHT, xml);
             int tabsPaddingLeft = getDimensPixelSize(R.styleable.LandingStrip_tabsLeftPadding, DEFAULT_TABS_PADDING, xml);
             int tabsPaddingRight = getDimensPixelSize(R.styleable.LandingStrip_tabsRightPadding, DEFAULT_TABS_PADDING, xml);
 
-            return new Attributes(tabLayoutId, indicatorColour, indicatorHeight, tabsPaddingLeft, tabsPaddingRight);
+            return new Attributes(indicatorColour, indicatorHeight, tabsPaddingLeft, tabsPaddingRight);
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/Attributes.java]: Can we haz `en-US` please?

File: core/src/main/java/com/novoda/landingstrip/MutableState.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package com.novoda.landingstrip;
+
+class MutableState implements State {
+
+    private float pagePositionOffset;
+    private int position;
+    private int fastForwardPosition;
+
+    public static MutableState newInstance() {
```
Comment (eduardb) [inline core/src/main/java/com/novoda/landingstrip/MutableState.java]: All these methods could be package-private :D

File: core/src/main/java/com/novoda/landingstrip/MutableState.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public void updatePosition(int position) {
+        this.position = position;
+    }
+
+    public void updateFastForwardPosition(int fastForwardPosition) {
+        this.fastForwardPosition = fastForwardPosition;
+    }
+
+    @Override
+    public float offset() {
```
Comment (eduardb) [inline core/src/main/java/com/novoda/landingstrip/MutableState.java]: 💯 for no `get` in the method names

File: core/src/main/java/com/novoda/landingstrip/Scrollable.java [hunk 4]
```diff
@@ -1,6 +1,6 @@
 package com.novoda.landingstrip;
 
-public interface Scrollable {
```
Comment (eduardb) [inline core/src/main/java/com/novoda/landingstrip/Scrollable.java:3]: 👍

File: core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java [hunk 5]
```diff
@@ -4,23 +4,24 @@
...
     private final ScrollOffsetCalculator scrollOffsetCalculator;
     private final Scrollable scrollable;
     private final FastForwarder fastForwarder;
-    private final OnPageChangedListenerCollection onPageChangedListenerCollection;
 
     private boolean firstTimeAccessed = true;
 
-    ScrollingPageChangeListener(State state, TabsContainer tabsContainer, ScrollOffsetCalculator scrollOffsetCalculator,
-                                Scrollable scrollable, FastForwarder fastForwarder, OnPageChangedListenerCollection onPageChangedListenerCollection) {
+    ScrollingPageChangeListener(MutableState state,
```
Comment (eduardb) [inline core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java]: Parameters on new line? 😄

File: core/src/main/java/com/novoda/landingstrip/State.java [hunk 6]
```diff
@@ -1,43 +1,10 @@
 package com.novoda.landingstrip;
 
-class State {
+interface State {
```
Comment (eduardb) [inline core/src/main/java/com/novoda/landingstrip/State.java:3]: 👍

File: core/src/main/java/com/novoda/landingstrip/TabsContainerView.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+public class TabsContainerView extends LinearLayout {
+
+    public TabsContainerView(Context context) {
+        super(context);
+        super.setOrientation(HORIZONTAL);
+    }
+
+    public void setActivated(int position) {
+        for (int index = 0; index < getChildCount(); index++) {
```
Comment (eduardb) [inline core/src/main/java/com/novoda/landingstrip/TabsContainerView.java:15]: Would it be worth having a reference to the last activated view instead of iterating over all the children? Or would that be error-prone?

File: README.md [hunk 8]
```diff
@@ -50,16 +50,41 @@ The tab strip which will contain the tab items
...
+
+    @Override
+    protected TextView createView(ViewGroup parent, int position) {
+        return (TextView) getLayoutInflater().inflate(R.layout.tab_simple_text, parent, false);
+    }
+
+    @Override
+    protected void bindView(TextView view, final int position) {
+        view.setText(titles[position]);
+        view.setOnClickListener(new View.OnClickListener() {
```
Comment (rock3r) [inline README.md]: I would argue `TabLayout`'s [`OnTabSelectedListener`](https://github.com/material-components/material-components-android/blob/master/lib/src/android/support/design/widget/TabLayout.java#L213-L236) is a better approach given tab selection is a very, very common scenario that the library should be able to handle by itself.

Another advantage of having a standardised "item selected" listener mechanism is that allows us to provide built-in but optional behaviour such as [forwarding tab selection to a `ViewPager`](https://github.com/material-components/material-components-android/blob/master/lib/src/android/support/design/widget/TabLayout.java#L2184-L2205) or [observing data changes in an adapter](https://github.com/material-components/material-components-android/blob/master/lib/src/android/support/design/widget/TabLayout.java#L2207-L2219).

No magic is great but optional built-in behaviour is great too for the users :) Those are trivial things to do but there's no point in not providing pluggable components for common patterns in this case. If users don't want to use them they're free not to, and if they want to do those things we have them ready for them to use and extend. I'm not saying we should attach those by default; I'm saying let's have a mechanism to make it easier to do.

**Review by eduardb — COMMENTED** [review]
I really like having an adapter for the tabs. But could we take this a step further? What I mean is that could we make use of the adapter to instantiate views on demand? Maybe even use a RecyclerView for the tabs? I imagine that inflating all the tabs in one go can be troublesome if the number of tabs is huge (not the common use case, but I just saw in an app a pager that has 90 tabs 😱).
Also, with this change, we are giving more work to the users of the library. Which is not necessarily bad, but couldn't we include in the library a basic adapter for standard material tabs, and/or one that takes in the constructor a layout ID (to seamlessly replace the previous `tabLayoutId`) 😄

File: core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java (hunk 5 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java]: would mean reformatting the whole lib 😨 would prefer not to! 😸

Comment (eduardb) [reply core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java]: Not necessarily. This is a line that you touched anyway, why not change it to better? Improvement can come in small steps 😄

File: core/src/main/java/com/novoda/landingstrip/TabsContainerView.java (hunk 7 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/TabsContainerView.java:15]: This logic is/was the existing but you make a good point, I feel like this is less error prone or is at least able to correct errors if any occur although is less efficient.

As far as I'm aware only the library internals are setting activated state so we could move away from iterating all the children, would prefer to do it in another PR though?

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java [hunk 9]
```diff
@@ -4,106 +4,99 @@
...
     }
 
-    public void addOnPageChangeListener(ViewPager.OnPageChangeListener onPageChangeListener) {
-        onPageChangeListenerCollection.add(onPageChangeListener);
-    }
+    @Override
+    protected void onFinishInflate() {
+        super.onFinishInflate();
+
+        this.tabsContainerView = new TabsContainerView(getContext());
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/LandingStrip.java:41]: Why not inflating it from XML?

File: core/src/main/java/com/novoda/landingstrip/TabsContainerView.java (hunk 7 above)
Comment (eduardb) [reply core/src/main/java/com/novoda/landingstrip/TabsContainerView.java:15]: I'm fine either way, but I would prefer some rigorous testing before such a change. Especially around configuration changes. But a different PR sounds better, so we can isolate the change.

**Comment by ouchadam** [comment]
@eduardb We've been thinking about moving to recyclerview, this change is a step along the way since the public apis should be identical!

I spoke with @ataulm a lot about whether to include some default adapters and decided against it as the `demo` provides an example already which should cover most usecases and we can always add more examples to the wiki/readme if needed. I agree that the migration from v1 to v2 will be a bit painful though 😢

**Comment by rock3r** [comment]
On the "recycling" bit, @eduardb I think it could be OK for now to leave it as a non-recycling behaviour to keep things simple. I doubt that will be needed in general as rarely you have more than 10-20 tabs in most cases (an alphabetic directory with one tab per letter is already annoying at 26-ish tabs). In addition, the tabs' views are bound to be quite easy. So I think we're ok for now. I don't know if it would be worth having this library in its current incarnation at all if your tabs were simply to be a synchronised-scrolling `RecyclerView` with an horizontal list layout manager...

File: core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java (hunk 5 above)
Comment (ataulm) [reply core/src/main/java/com/novoda/landingstrip/ScrollingPageChangeListener.java]: pfft how big is the lib :P

**Comment by ataulm** [comment]
>LandingStrip no longer automagically sets a click listener on the tab views, the client must do this manually AND call LandingStrip.setCurrentItem(position), this is also required if and when calling ViewPager.setCurrentItem(position) manually.

@ouchadam why:

>this is also required if and when calling ViewPager.setCurrentItem(position) manually.

? If you call setCurrentItem, doesn't it trigger the page change listener?

**Comment by ouchadam** [comment]
@ataulm maybe I need to reword, it's only needed if you want to instantly set the indicator position to the new current item (which is `v1` and the `design tab layout` behaviour) otherwise the indicator will scroll into the position.

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java [hunk 10]
```diff
@@ -134,79 +127,18 @@ public void scrollTo(int x, int y) {
     }
 
     @Override
-    protected void onAttachedToWindow() {
-        super.onAttachedToWindow();
-        pagerAdapterObserver.reregister();
+    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
+        scrollingPageChangeListener.onPageScrolled(position, positionOffset, positionOffsetPixels);
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: Why isn't the `ScrollingPageChangeListener` implementing the interface itself given it de facto is already? That way we can spare the purely delegating four methods on here.

File: core/src/main/java/com/novoda/landingstrip/LandingStripAdapter.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        if (listener != null) {
+            listener.onNotifyItemChanged(this, position);
+        }
+    }
+
+    protected abstract int getCount();
+
+    interface Listener {
+
+        <T extends View> void onNotifyDataSetChanged(LandingStripAdapter<T> adapter);
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/LandingStripAdapter.java]: Why isn't `T` defined as a type parameter on the interface itself? The two methods can currently pass in two different adapters with two different types in theory, it's almost no different than making them pass an adapter of `View`

File: core/src/main/java/com/novoda/landingstrip/Notifier.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    Notifier(TabsContainerView tabsContainerView) {
+        this.tabsContainerView = tabsContainerView;
+    }
+
+    @Override
+    public <T extends View> void onNotifyDataSetChanged(LandingStripAdapter<T> adapter) {
+        recreateAndBindTabs(adapter);
+    }
+
+    private <T extends View> void recreateAndBindTabs(LandingStripAdapter<T> adapter) {
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/Notifier.java]: I understand the purpose of documenting what it does but even if it were inlined it'd be quite clear...

Comment (rock3r) [reply core/src/main/java/com/novoda/landingstrip/Notifier.java]: Also, in the future we should be [recycling the views](https://github.com/rock3r/squanchy/blob/develop/app/src/main/java/net/squanchy/support/widget/SpeakerView.java#L79) if possible. We might need some sort of contract such as view types in the adapter, but that shouldn't be too hard. Currently this operation can be rather expensive if you have several tabs.

File: core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java [hunk 13]
```diff
@@ -1,26 +1,29 @@
...
 
-    ScrollOffsetCalculator(TabsContainer tabsContainer) {
+    ScrollOffsetCalculator(View rootView, TabsContainerView tabsContainer) {
+        this.rootView = rootView;
         this.tabsContainer = tabsContainer;
     }
 
     int calculateScrollOffset(int position, float pagerOffset) {
-        if (!tabsContainer.hasTabAt(position)) {
+        if (!hasTabAt(position)) {
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java]: I'm quite sure this should be an `IndexOutOfBoundsException` :P

File: core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java [hunk 14]
```diff
@@ -32,20 +35,20 @@ private int roundToInt(float input) {
...
+    private int getHorizontalScrollOffset(int position, float pagerOffset) {
+        int tabWidth = tabsContainer.getChildAt(position).getWidth();
         return Math.round(pagerOffset * tabWidth);
     }
 
     private float getNextTabDelta(int position, float pagerOffset, View tabForPosition) {
-        if (tabsContainer.hasTabAt(position + 1)) {
-            return (((tabsContainer.getTabAt(position + 1).getWidth()) - tabForPosition.getWidth()) * pagerOffset) * HALF_MULTIPLIER;
+        if (hasTabAt(position + 1)) {
+            return (((tabsContainer.getChildAt(position + 1).getWidth()) - tabForPosition.getWidth()) * pagerOffset) * HALF_MULTIPLIER;
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java]: Can we extract things to make this more readable?

File: core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java [hunk 15]
```diff
@@ -1,26 +1,29 @@
...
-        if (!tabsContainer.hasTabAt(position)) {
+        if (!hasTabAt(position)) {
             return 0;
         } else {
-            View tabForPosition = tabsContainer.getTabAt(position);
+            View tabForPosition = tabsContainer.getChildAt(position);
 
-            float tabStartX = tabForPosition.getLeft() + getHorizontalScrollOffset(tabForPosition, pagerOffset);
-            float viewMiddleOffset = getTabParentWidth() * HALF_MULTIPLIER;
+            float tabStartX = tabForPosition.getLeft() + getHorizontalScrollOffset(position, pagerOffset);
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java]: Why this change? You're doing `tabForPosition = tabsContainer.getChildAt(position)` twice now. Not a big perf concern, I just don't follow the reasoning

File: core/src/main/java/com/novoda/landingstrip/State.java [hunk 16]
```diff
@@ -1,43 +1,10 @@
...
-    public int getPosition() {
-        return position;
-    }
-
-    public int getFastForwardPosition() {
-        return fastForwardPosition;
-    }
+    float offset();
 
+    int fastForwardPosition();
```
Comment (rock3r) [inline core/src/main/java/com/novoda/landingstrip/State.java]: Missing whiteline after

File: core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java (hunk 13 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java]: nah @joetimmins fixed this with `return position < tabsContainer.getChildCount();`

File: core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java (hunk 15 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java]: ah bad merge, will fix

File: core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java (hunk 13 above)
Comment (rock3r) [reply core/src/main/java/com/novoda/landingstrip/ScrollOffsetCalculator.java]: As in the library should never let this happen. Swallow != fix :P

File: README.md (hunk 8 above)
Comment (ataulm) [reply README.md]: @rock3r what's expected behavior if user sets a click listener on the view and they provide a tabSelectedListener?

Comment (ouchadam) [reply README.md]: I agree and had tried but there's a few issues with this approach:
- The library doesn't supply a custom view for each tab, maybe it should?
- Exposing a `LandingStrip.setOnTabClickListener` means that `bindView` could override this `Listener`
- The indicator will become out of sync if the `OnTabClick` is calling `LandingStrip.setCurrentItem` but manually calling `ViewPager.setCurrentItem` isn't. 
The only way to enforce consistency would be to do the `ViewPager.setCurrentItem` via the `LandingStrip` and reintroduce `attach` which the original aim was to decouple 😢.  

I'm all ears though! 👻

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java (hunk 10 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: not sure I follow, it is implementing the interface but the implementation is extracted?

Or do you mean exposing the  `ScrollingPageChangeListener` as a public api instead of setting the `LandingStrip` as a page change listener?

File: core/src/main/java/com/novoda/landingstrip/LandingStripAdapter.java (hunk 11 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/LandingStripAdapter.java]: Because `T` is a client concept, the library doesn't know T at the point of creating the `Listener` https://github.com/novoda/landing-strip/pull/32/files/0d559730914d7e4f1e3278979c5845fe1d6bdb05#diff-03f70e68af4374fc21ddd30d94ca1309R57

`T` is only known once `setAdapter` has been called.

Perhaps it would be better to create the `listener<T>` in `setAdapter` instead of allowing dynamic types

File: README.md (hunk 8 above)
Comment (rock3r) [reply README.md]: >- The library doesn't supply a custom view for each tab, maybe it should?

I'm not sure I follow this, isn't the tab view created by the (user supplied) adapter? Then they can be literally anything.

> - Exposing a `LandingStrip.setOnTabClickListener` means that `bindView` could override this `Listener`

Just set it after calling `bindView()` and maybe error out if there is a click listener set otherwise

> - The indicator will become out of sync if the `OnTabClick` is calling `LandingStrip.setCurrentItem` but manually calling `ViewPager.setCurrentItem` isn't. 
The only way to enforce consistency would be to do the `ViewPager.setCurrentItem` via the `LandingStrip` and reintroduce `attach` which the original aim was to decouple 😢.  

If you have an optional built-in facility for `ViewPagers` this should be fine; if the user wants to hurt themselves and put their hands into the inner workings then it's kind of their problem I suppose... unless we can find a better way.

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java (hunk 10 above)
Comment (rock3r) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: The latter, with the difference that if `ScrollingPageChangeListener` implements the interface you don't need to expose it for the purpose of setting a scroll listener, you can pass around the interface (and not set `this` as listener but the scrolling page listener directly, which is the point I was trying to make)

**Comment by ouchadam** [comment]
closing to address comments

File: core/src/main/java/com/novoda/landingstrip/LandingStripAdapter.java (hunk 11 above)
Comment (rock3r) [reply core/src/main/java/com/novoda/landingstrip/LandingStripAdapter.java]: >Perhaps it would be better to create the `listener<T>` in `setAdapter` instead of allowing dynamic types

Yeah I'd rather do that, seems more correct given that is the scope where you know the types

File: core/src/main/java/com/novoda/landingstrip/Notifier.java (hunk 12 above)
Comment (ouchadam) [reply core/src/main/java/com/novoda/landingstrip/Notifier.java]: I totally agree but think the recycling effort would be more appreciated here https://github.com/novoda/view-pager-adapter

Comment (rock3r) [reply core/src/main/java/com/novoda/landingstrip/Notifier.java]: Yeah it's ok for now, I was mostly thinking out loud there :)

**Comment by ouchadam** [comment]
comments addressed in 
- #34 typing notifier
- #35 offset calculator tidy up,
- #36 adds a basic adapter implementation
- #37 fixes the initial tab activation state not being set

**Review by rock3r — APPROVED** [review]

**Review by ataulm — APPROVED** [review]

**Review by eduardb — APPROVED** [review]
Cool!
